README.md
# Qh

Easy Rails-style query helper for iex. 

```elixir
use Qh

Qh.configure(app: :my_app)

q User.where(age > 20).limit(10).all
[%MyApp.User{age: 22, name: "Bob"}, ...]
```

## Installation

Add `qh` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:qh, "~> 0.2"}
  ]
end
```

Add to your `.iex.exs`:

```elixir
use Qh

Qh.configure(app: :my_app)
```

Make sure to change `:my_app` to your otp app name that holds the schema's and repo's.

## Usage

Most `Ecto.Query` and `Ecto.Repo` functions are supported.

 - Call by chaining query/repo functions
 - Specifying bindings is optional if you're fetching for a single schema. 
 - Pass a condition as binary to use custom fragments
 - Some added aliases for convenience

### Query examples

```elixir
# First/last
q User.first
q User.first(10)
q User.last
q User.last(1)

# Custom order
q User.order(name).last(3)
q User.order(name: :asc, age: :desc).last
q User.order("lower(?)", name).last

# Conditions
q User.where(age > 20 and age <= 30).count
q User.where(age > 20 and age <= 30).limit(10).all
q User.where(age > 20 or name == "Bob").all
q User.where(age > 20 and (name == "Bob" or name == "Anna")).all
q User.where(age: 20, name: "Bob").count
q User.where("nicknames && ?", ["Bobby", "Bobi"]).count
q User.where("? = ANY(?)", age, [20, 30, 40]).count

# Opional binding
q User.where([u], u.age > 20 and u.age <= 30).count

# Find
q User.get!(21)
# or
q User.find(21)

# Alias for where(...).first
q User.find_by(name: "Bob Foo")
q User.find_by(name == "Bob" or name == "Anna")

# Aggregations
q User.group_by("length(name)").count
q User.group_by(name).avg(age)

# Select stats
q User.select(count(), avg(age), min(age), max(age)).all

# Aggregate stats grouped by column
q User.group_by(name).aggr(%{count: count(), avg: avg(age), min: min(age), max: max(age)})

# Count number of messages per user
q User.left_join(:messages).group_by(id).count([u, m], m.id)

# Grab only users that have messages
q User.distinct(id).join(:messages).all

# Custom join logic
q User.join([u], u in MyApp.Messages, on: u.id == m.sent_by_id, as: :m)
```

## Configuration

 - `app`: App name that is used for infering the schema namespace and repo
 - `app_mod`: You can also set the app module directly, instead of the app name
 - `repo`: Set the repo module directly, in case of a non-default repo

Set via config:

```elixir
config :qh, app: :my_app, repo: MyApp.AlternateRepo
```

Set via configure:

```elixir
Qh.configure(app: :my_app, repo: MyApp.AlternateRepo)
```

Pass in options:

```elixir
user = q(User.find(21), app: :my_app, repo: MyApp.AlternateRepo)
```


CHANGELOG.md
# Changelog for Qh

## v0.2.0

- Support for most `Ecto.Query` functions
- Support other `Ecto.Repo` functions
- Allow optional bindings
- Added `aggr` to allow for grouped aggregations 

## v0.1.0

Initial version


test/q_test.exs
defmodule QTest do
  use Qh.DataCase, async: true

  use Qh

  setup do
    bob = QTest.Repo.insert!(%QTest.User{name: "Bob", age: 22, nicknames: ["Bobi", "Bobby"]})
    anna = QTest.Repo.insert!(%QTest.User{name: "Anna", age: 21, nicknames: ["Ann"]})
    _james = QTest.Repo.insert!(%QTest.User{name: "James", age: 65})
    _john = QTest.Repo.insert!(%QTest.User{name: "John", age: 40})

    QTest.Repo.insert!(%QTest.Message{user_id: bob.id, message: "Message to bob 1", likes: 5})
    QTest.Repo.insert!(%QTest.Message{user_id: bob.id, message: "Message to bob 2", likes: 4})
    QTest.Repo.insert!(%QTest.Message{user_id: anna.id, message: "Message to anna 1", likes: 2})
    QTest.Repo.insert!(%QTest.Message{user_id: anna.id, message: "Message to anna 2", likes: 3})
    QTest.Repo.insert!(%QTest.Message{user_id: anna.id, message: "Message to anna 3", likes: 2})

    :ok
  end

  test "first" do
    assert %QTest.User{name: "Bob"} = q(User.first())
  end

  test "last" do
    assert %QTest.User{name: "John"} = q(User.last())
  end

  test "first n" do
    assert [%QTest.User{name: "Bob"}, %QTest.User{name: "Anna"}] = q(User.first(2))
  end

  test "last n" do
    assert [%QTest.User{name: "James"}, %QTest.User{name: "John"}] = q(User.last(2))
  end

  test "count" do
    assert 4 = q(User.count())
  end

  test "select" do
    assert ["Bob", "Anna", "James", "John"] = q(User.select(name).all())
  end

  test "select multiple" do
    assert [["Bob", 22], ["Anna", 21], ["James", 65], ["John", 40]] =
             q(User.select(name, age).all())
  end

  test "select custom struct" do
    assert [%{name: "Bob"}, %{name: "Anna"}, %{name: "James"}, %{name: "John"}] =
             q(User.select(%{name: name}).all())
  end

  test "select with binding" do
    assert ["Bob", "Anna", "James", "John"] = q(User.select([u], u.name).all())
  end

  test "select merge" do
    assert [
             %{age: 22, name: "Bob"},
             %{age: 21, name: "Anna"},
             %{age: 65, name: "James"},
             %{age: 40, name: "John"}
           ] ==
             q(
               User.order(id).select(%{}).select_merge(%{name: name}).select_merge(%{age: age}).all
             )
  end

  test "where" do
    assert 1 = q(User.where(age > 25 and age < 50).count)
  end

  test "where with binding" do
    assert 1 = q(User.where([u], u.age > 25 and u.age < 50).count)
  end

  test "where or" do
    assert 2 = q(User.where(name == "Bob" or name == "Anna").count)
  end

  test "where nested" do
    assert 2 =
             q(
               User.where(age < 50 and (name == "Bob" or name == "Anna" or name == "James")).count
             )
  end

  test "pinned where" do
    name = "Bob"
    assert 1 = q(User.where(name: ^name).count)
  end

  test "keyword where" do
    assert 1 = q(User.where(age: 22, name: "Bob").count)
  end

  test "fragment where" do
    assert 2 = q(User.where("nicknames && ?", ["Bobi", "Ann"]).count)
  end

  test "fragment where 2" do
    assert 2 = q(User.where("name = ANY(?)", ["Bob", "Anna"]).count)
  end

  test "fragment where with field" do
    assert 1 = q(User.where("? && ?", nicknames, ["Bobi", "Bobby"]).count)
  end

  test "having" do
    assert [5] =
             q(User.group_by(name).select("length(?)", name).having("length(?) > ?", name, 4).all)
  end

  test "find_by" do
    assert %{name: "Bob"} = q(User.find_by(name: "Bob"))
  end

  test "find_by non existing" do
    assert nil == q(User.find_by(name: "Foo"))
  end

  test "find_by complex" do
    assert %{name: "John"} = q(User.find_by(age > 25 and age < 50))
  end

  test "find_by fragment" do
    assert %{name: "James"} = q(User.find_by("age = ?", 65))
  end

  test "order" do
    assert %{age: 21} = q(User.order(age).first)
  end

  test "order desc" do
    assert %{age: 65} = q(User.order(age: :desc).first)
  end

  test "order multiple" do
    assert %{age: 65} = q(User.order(age: :desc, name: :asc).first)
  end

  test "order by fragment" do
    assert %{name: "Anna"} = q(User.order("lower(?)", name).first)
  end

  test "group_by count" do
    assert [{3, 1}, {5, 1}, {4, 2}] = q(User.group_by("length(name)").count)
  end

  test "group_by avg" do
    assert [{3, avg} | _] = q(User.group_by("length(name)").avg(age))
    assert Decimal.equal?(avg, Decimal.from_float(22.0))
  end

  test "aggr multiple" do
    assert [{3, {1, 22}}, {5, {1, 65}} | _] =
             q(User.group_by("length(name)").aggr({count(), min(age)}))
  end

  test "aggr without group_by" do
    assert {4, 21} = q(User.aggr({count(), min(age)}))
  end

  test "limit" do
    assert 2 = length(q(User.limit(2).all))
  end

  test "query" do
    assert %Ecto.Query{} = q(User.query())
  end

  test "preload" do
    assert %{messages: messages} = q(User.preload(:messages).first)
    assert length(messages) == 2
  end

  test "preload custom query" do
    messages_query = q(Message.order(likes))

    assert %{messages: [%{likes: 4}, %{likes: 5}]} =
             q(User.preload(messages: ^messages_query).first)
  end

  test "preload custom query direct" do
    assert %{messages: [%{likes: 4}, %{likes: 5}]} =
             q(User.preload(messages: ^q(Message.order(likes))).first)
  end

  test "reverse_order" do
    assert [%{name: "John"} | _] = q(User.order(id).reverse_order.all)
  end

  test "except_all" do
    oldest_2 = q(User.order(age: :desc).limit(2))

    assert [%{name: "Bob"}, %{name: "Anna"}] =
             q(User.except_all(^oldest_2).all) |> Enum.sort_by(fn u -> u.id end)
  end

  test "except" do
    oldest_2 = q(User.order(age: :desc).limit(2))

    assert [%{name: "Bob"}, %{name: "Anna"}] =
             q(User.except(^oldest_2).all) |> Enum.sort_by(fn u -> u.id end)
  end

  test "exclude" do
    assert %{name: "Bob"} = q(User.select(age).exclude(:select).first)
  end

  test "intersect" do
    not_bob = q(User.where(name != "Bob"))
    not_anna = q(User.where(name != "Anna"))

    assert [%{name: "James"}, %{name: "John"}] =
             q(not_bob.intersect(^not_anna).all) |> Enum.sort_by(fn u -> u.id end)
  end

  test "intersect all" do
    not_bob = q(User.where(name != "Bob"))
    not_anna = q(User.where(name != "Anna"))

    assert [%{name: "James"}, %{name: "John"}] =
             q(not_bob.intersect_all(^not_anna).all) |> Enum.sort_by(fn u -> u.id end)
  end

  test "exists?" do
    assert true == q(User.where(name: "Bob").exists?)
    assert false == q(User.where(name: "Noone").exists?)
  end

  test "one" do
    assert %{name: "Bob"} = q(User.where(name: "Bob").one)
  end

  test "one returns nil for empty" do
    assert nil == q(User.where(name: "Noone").one)
  end

  test "one raises for multiple" do
    assert_raise Ecto.MultipleResultsError, fn -> q(User.one()) end
  end

  test "stream" do
    assert %Stream{} = q(User.stream())
  end

  test "join" do
    assert 2 = q(User.distinct(id).join(:messages).all) |> length()
  end

  test "custom join" do
    assert 2 =
             q(User.distinct(id).join([u], m in assoc(u, :messages), on: u.id == m.user_id).all)
             |> length()
  end

  # modify

  test "new" do
    assert %QTest.User{} = q(User.new())
  end

  test "new with params" do
    assert %QTest.User{name: "Alice", age: 25} = q(User.new(name: "Alice", age: 25))
  end

  test "save new" do
    user = q(User.new(name: "Alice", age: 25))
    assert {:ok, _user} = Qh.Repo.save(user)
    assert 5 == q(User.count())
  end

  test "save existing" do
    user = q(User.first())
    assert {:ok, _user} = Qh.Repo.assign(user, age: 23) |> Qh.Repo.save()
    assert q(User.find_by(name: ^user.name)).age == 23
  end

  test "save validation failure" do
    assert {:error, [name: {"can't be blank", [validation: :required]}]} =
             Qh.Repo.save(%QTest.User{})
  end

  test "update" do
    user = q(User.first())
    assert {:ok, _user} = Qh.Repo.update(user, age: 23)
    assert q(User.find_by(name: ^user.name)).age == 23
  end

  test "update non existing" do
    assert_raise Ecto.NoPrimaryKeyValueError, fn ->
      Qh.Repo.update(%QTest.User{name: "Test"}, age: 23)
    end
  end

  test "delete!" do
    user = q(User.first())
    Qh.Repo.delete!(user)
    assert 3 == q(User.count())
  end
end


test/support/repo.ex
defmodule QTest.Repo do
  use Ecto.Repo, otp_app: :qh, adapter: Ecto.Adapters.Postgres
end


test/support/user.ex
defmodule QTest.User do
  use Ecto.Schema
  import Ecto.Changeset

  schema "users" do
    field(:name, :string)
    field(:age, :integer)
    field(:nicknames, {:array, :string})
    has_many :messages, QTest.Message
    timestamps()
  end

  def changeset(user \\ %__MODULE__{}, params) do
    user
    |> cast(params, __schema__(:fields))
    |> validate_required([:name])
  end
end


mix.exs
defmodule Qh.MixProject do
  use Mix.Project

  def project do
    [
      app: :qh,
      version: "0.2.0",
      elixir: "~> 1.10",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      package: package(),
      description: description(),
      docs: docs(),
      deps: deps()
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp deps do
    [
      {:ecto, "~> 3.0"},
      {:ecto_sql, "~> 3.7", only: :test},
      {:postgrex, ">= 0.0.0", only: :test},
      {:ex_doc, "~> 0.26", only: :dev, runtime: false}
    ]
  end

  def description do
    """
    Ecto query helper for iex (Rails style)
    """
  end

  defp package do
    [
      name: :qh,
      files: ["lib", "mix.exs", "README*", "LICENSE*", "CHANGELOG*"],
      maintainers: ["Robin Fidder"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/fidr/qh"}
    ]
  end

  defp docs do
    [
      name: "Q helper",
      source_url: "https://github.com/fidr/qh",
      homepage_url: "https://github.com/fidr/qh",
      main: "readme",
      extras: ["README.md"]
    ]
  end
end


lib/qh/expr.ex
defmodule Qh.Expr do
  def maybe_deep_prefix_binding(skip_cond, tree) do
    if skip_cond do
      tree
    else
      deep_prefix_binding(tree)
    end
  end

  def deep_prefix_binding(tree, prefix \\ {:t, [], nil}) do
    conditional_prewalk(tree, fn
      {:^, _, _} = node ->
        {false, node}

      {field, _, nil} ->
        {false, {{:., [], [prefix, field]}, [no_parens: true], []}}

      node ->
        {true, node}
    end)
  end

  def conditional_prewalk(tree, fun) do
    apply_and_maybe_descend(tree, fun)
  end

  defp do_prewalk_cond({form, meta, args}, fun) when is_atom(form) do
    {form, meta, do_prewalk_cond_args(args, fun)}
  end

  defp do_prewalk_cond({form, meta, args}, fun) do
    {apply_and_maybe_descend(form, fun), meta, do_prewalk_cond_args(args, fun)}
  end

  defp do_prewalk_cond({left, right}, fun) do
    {apply_and_maybe_descend(left, fun), apply_and_maybe_descend(right, fun)}
  end

  defp do_prewalk_cond(list, fun) when is_list(list) do
    do_prewalk_cond_args(list, fun)
  end

  defp do_prewalk_cond(x, _fun) do
    x
  end

  defp do_prewalk_cond_args(args, _fun) when is_atom(args) do
    args
  end

  defp do_prewalk_cond_args(args, fun) when is_list(args) do
    Enum.map(args, fn node -> apply_and_maybe_descend(node, fun) end)
  end

  defp apply_and_maybe_descend(node, fun) do
    case fun.(node) do
      {true, node} ->
        do_prewalk_cond(node, fun)

      {false, node} ->
        node
    end
  end
end

lib/qh/repo.ex
defmodule Qh.Repo do
  @doc """
  Update fields in a struct

  Examples:

      user = %MyApp.User{name: nil, age: nil}
      user = Qh.assign(user, name: "Bob", age: 21)

      # or initialize a new record
      user = Qh.assign(:user, name: "Alice", age: 22)
  """
  def assign(schema, params, opts \\ [])

  def assign(%_{} = struct, params, _opts) do
    struct(struct, params)
  end

  def assign(schema, params, opts) do
    struct(Qh.lookup_schema(schema, opts), params)
  end

  @doc """
  Save a struct to the database

  Inserts a new record if no primary key is set
  or no record exists with the current primary key.

  Performs a get and an update if the record already exists.

  Returns `{:ok, struct}` on success or `{:error, validation_errors}` on failure.

  Examples:
      user = %MyApp.User{id: nil, name: "Bob", age: 21}
      {:ok, user} = Qh.save(user)

  """
  def save(%schema{} = struct, opts \\ []) do
    primary_key = schema.__schema__(:primary_key)
    primary_key_values = Map.take(struct, primary_key)
    primary_values = Map.values(primary_key_values)

    if Enum.all?(primary_values, &is_nil/1) do
      # no primary key set, create new new
      params = Map.from_struct(struct)
      changeset = schema.changeset(struct(schema), params)

      Qh.repo(opts).insert(changeset)
      |> handle_repo_result()
    else
      case Qh.repo(opts).get_by(schema, primary_key_values) do
        nil ->
          # no record found, create new
          params = Map.from_struct(struct)
          changeset = schema.changeset(struct(schema), params)

          Qh.repo(opts).insert(changeset)
          |> handle_repo_result()

        current ->
          # record found, update
          params = Map.from_struct(struct)
          changeset = schema.changeset(current, params)

          Qh.repo(opts).update(changeset)
          |> handle_repo_result()
      end
    end
  end

  @doc """
  Save a struct to the database

  Similar to `save/1` but will raise on failures and return only the
  struct on success.

  Examples:
      user = %MyApp.User{id: nil, name: "Bob", age: 21}
      user = Qh.save!(user)

  """
  def save!(%_schema{} = struct, opts \\ []) do
    save(struct, opts) |> ok_or_raise()
  end

  @doc """
  Performs a database update for a struct and params

  Returns `{:ok, struct}` on success or `{:error, validation_errors}` on failure.

  Examples:
      user = %MyApp.User{id: 2, name: "Bob", age: 21}
      {:ok, user} = Qh.update(user, age: 22)

  """
  def update(%schema{} = struct, params, opts \\ []) do
    changeset = schema.changeset(struct, Map.new(params))

    Qh.repo(opts).update(changeset)
    |> handle_repo_result()
  end

  @doc """
  Save a struct to the database

  Similar to `update/1` but will raise on failures and return only the
  struct on success.

  Examples:
      user = %MyApp.User{id: 2, name: "Bob", age: 21}
      user = Qh.update!(user, age: 22)

  """
  def update!(%_schema{} = struct, params, opts \\ []) do
    update(struct, params, opts) |> ok_or_raise()
  end

  @doc """
  Delete a struct in the database

  Will raise if the record doesn't exists

  Examples:
      user = %MyApp.User{id: 2, name: "Bob", age: 21}
      Qh.delete!(user)

  """
  def delete!(%_schema{} = struct, opts \\ []) do
    Qh.repo(opts).delete!(struct)
  end


  defp handle_repo_result({:ok, struct}) do
    {:ok, struct}
  end

  defp handle_repo_result({:error, changeset}) do
    {:error, changeset.errors}
  end

  defp ok_or_raise({:ok, struct}) do
    struct
  end

  defp ok_or_raise({:error, error}) do
    raise inspect(error)
  end
end

lib/qh.ex
defmodule Qh do
  @moduledoc """
  Query helper for iex

  `.iex.exs`:

  ```
  use Qh

  Qh.configure(app: :my_app)
  ```

  ## Example:

      iex>q User.where(age > 20 and age < 50).order(name).first
      %MyApp.User%{id: 123, age: 21, name: "Anna"}
  """

  defmacro __using__(_) do
    quote do
      require Qh
      import Qh, only: [q: 1, q: 2]
    end
  end

  defmacro q(query, opts \\ []) do
    Qh.query(query, opts, __CALLER__)
  end


  def configure(opts) do
    Enum.each(opts, fn {k, v} ->
      Application.put_env(:qh, k, v)
    end)
  end

  @doc """
  Run a query

  Examples:

      # First/last
      q User.first
      q User.first(10)
      q User.last
      q User.last(1)

      # Custom order
      q User.order(name).last(3)
      q User.order(name: :asc, age: :desc).last
      q User.order("lower(?)", name).last

      # Conditions
      q User.where(age > 20 and age <= 30).count
      q User.where(age > 20 and age <= 30).limit(10).all
      q User.where(age > 20 or name == "Bob").all
      q User.where(age > 20 and (name == "Bob" or name == "Anna")).all
      q User.where(age: 20, name: "Bob").count
      q User.where("nicknames && ?", ["Bobby", "Bobi"]).count
      q User.where("? = ANY(?)", age, [20, 30, 40]).count

      # Opional binding
      q User.where([u], u.age > 20 and u.age <= 30).count

      # Find
      q User.get!(21)
      # or
      q User.find(21)

      # Alias for where(...).first
      q User.find_by(name: "Bob Foo")
      q User.find_by(name == "Bob" or name == "Anna")

      # Aggregations
      q User.group_by("length(name)").count
      q User.group_by(name).avg(age)

      # Select stats
      q User.select(count(), avg(age), min(age), max(age)).all

      # Aggregate stats grouped by column
      q User.group_by(name).aggr(%{count: count(), avg: avg(age), min: min(age), max: max(age)})

      # Count number of messages per user
      q User.left_join(:messages).group_by(id).count([u, m], m.id)

      # Grab only users that have messages
      q User.distinct(id).join(:messages).all

      # Custom join logic
      q User.join([u], u in MyApp.Messages, on: u.id == m.sent_by_id, as: :m)

  """
  def query(q, opts \\ [], caller \\ __ENV__) do
    [schema | rest] = unwrap_nested(q)

    code =
      if is_atom(schema) && !first_uppercase?(schema) do
        quote do
          require Ecto.Query
          query = unquote({schema, [if_undefined: :apply], nil})
        end
      else
        quote do
          require Ecto.Query
          query = Qh.lookup_schema(unquote(schema), unquote(opts))
        end
      end

    # aliases
    rest =
      Enum.flat_map(rest, fn
        {:order, _, params} ->
          [{:order_by, [], params}]

        {:find, _, params} ->
          [{:get, [], params}]

        {:find_by, _, params} ->
          [{:where, [], params}, {:first, [], []}]

        {fun, _, params} when fun in [:count, :sum, :avg, :min, :max] ->
          if first_param_binding?(params) do
            [binding | params] = params
            [{:aggr, [], [binding, {fun, [], params}]}]
          else
            [{:aggr, [], [{fun, [], params}]}]
          end

        other ->
          [other]
      end)

    code =
      [code] ++
        Enum.map(rest, fn
          # Ecto.Query functions /3 (with binding)
          {fun, _, params}
          when fun in [
                 :distinct,
                 :where,
                 :select,
                 :select_merge,
                 :group_by,
                 :having,
                 :lock,
                 :or_having,
                 :or_where,
                 :order_by,
                 :preload,
                 :windows,
                 :limit
               ] ->
            transform_with_binding(fun, params)

          # Ecto.Query functions /2
          {fun, _, params}
          when fun in [
                 :except,
                 :except_all,
                 :exclude,
                 :intersect,
                 :intersect_all,
                 :union,
                 :union_all
               ] ->
            quote do
              query = Ecto.Query.unquote(fun)(query, unquote_splicing(params))
            end

          # Ecto.Query functions /1
          {fun, _, []}
          when fun in [
                 :reverse_order
               ] ->
            quote do
              query = Ecto.Query.unquote(fun)(query)
            end

          # Repo functions /1
          {fun, _, []} when fun in [:all, :one, :stream, :exists?] ->
            quote do
              Qh.repo(unquote(opts)).unquote(fun)(query)
            end

          # Quick or custom join
          {join, _, params}
          when join in [
                 :join,
                 :inner_join,
                 :left_join,
                 :right_join,
                 :cross_join,
                 :full_join,
                 :inner_lateral_join,
                 :left_lateral_join
               ] ->
            transform_join(join, params)

          # Aggregrate respecting group_bys
          {:aggr, _, params} ->
            transform_aggr(params, caller, opts)

          # Return a queryable
          {:query, _, []} ->
            quote do
              query = Ecto.Queryable.to_query(query)
            end

          # First/last
          {:first, _, []} ->
            quote do
              Ecto.Query.limit(query, 1)
              |> Qh.default_primary_order()
              |> Qh.repo(unquote(opts)).one()
            end

          {:first, _, [n]} ->
            quote do
              Ecto.Query.limit(query, unquote(n))
              |> Qh.default_primary_order()
              |> Qh.repo(unquote(opts)).all()
            end

          {:last, _, []} ->
            quote do
              Ecto.Query.reverse_order(query)
              |> Ecto.Query.limit(1)
              |> Qh.default_primary_order()
              |> Qh.repo(unquote(opts)).one()
            end

          {:last, _, [n]} ->
            quote do
              Ecto.Query.reverse_order(query)
              |> Ecto.Query.limit(unquote(n))
              |> Qh.default_primary_order()
              |> Qh.repo(unquote(opts)).all()
              |> Enum.reverse()
            end

          {:new, _, []} ->
            quote do
              schema = Qh.get_schema(query)
              struct(schema)
            end

          {:new, _, [params]} ->
            quote do
              schema = Qh.get_schema(query)
              struct(schema, unquote(params))
            end
        end)

    code =
      quote do
        (unquote_splicing(code))
      end

    #IO.puts(Macro.to_string(code))

    code
  end

  # Transform

  def transform_join(fun, params) do
    type =
      case String.split(to_string(fun), "_", parts: 2) do
        ["join"] -> :inner
        [type, "join"] -> String.to_atom(type)
      end

    case params do
      [target] when is_atom(target) ->
        quote do
          query = Ecto.Query.join(query, unquote(type), [t], assoc(t, unquote(target)))
        end

      [target, as] when is_atom(target) and is_atom(as) ->
        quote do
          query =
            Ecto.Query.join(query, unquote(type), [t], assoc(t, unquote(target)),
              as: unquote(as)
            )
        end

      [target, {as, _, nil}] when is_atom(target) and is_atom(as) ->
        quote do
          query =
            Ecto.Query.join(query, unquote(type), [t], assoc(t, unquote(target)),
              as: unquote(as)
            )
        end

      params ->
        {binding_provided, binding, params} = Qh.split_optional_binding(params)

        params = Qh.Expr.maybe_deep_prefix_binding(binding_provided, params)

        quote do
          query =
            Ecto.Query.join(
              query,
              unquote(type),
              unquote(binding),
              unquote_splicing(params)
            )
        end
    end
  end

  def transform_aggr(original_params, caller, opts) do
    {binding_provided, binding, params} = Qh.split_optional_binding(original_params)

    {_, binding} =
      Ecto.Query.Builder.escape_binding(quote(do: %Ecto.Query{}), binding, caller)

    prefixed_params = unwrap_single(params)

    prefixed_params = Qh.Expr.maybe_deep_prefix_binding(binding_provided, prefixed_params)

    {prefixed_params, _take} =
      Ecto.Query.Builder.Select.escape(prefixed_params, binding, caller)

    call_select = transform_with_binding(:select, original_params)

    quote do
      query =
        if Qh.has_group_by?(query) do
          query
          |> Qh.group_by_and_aggregate(unquote(prefixed_params))
          |> Qh.repo(unquote(opts)).all()
        else
          unquote(call_select)
          |> Qh.repo(unquote(opts)).one()
        end
    end
  end

  def group_by_and_aggregate(query, target) do
    {first_group_by, joined_expr} =
      case query.group_bys do
        [group_by] ->
          {group_by, unwrap_single(group_by.expr)}

        [first | _] = list ->
          {first, Enum.map(list, fn group_by -> unwrap_single(group_by.expr) end)}
      end

    target = expr_to_funcs(target)

    expr = {:{}, [], [joined_expr, target]}

    select = %Ecto.Query.SelectExpr{
      expr: expr,
      file: first_group_by.file,
      line: first_group_by.line
    }

    Map.put(query, :select, select)
  end

  def transform_with_binding(fun, params) do
    {binding_provided, binding, params} = Qh.split_optional_binding(params)

    params = replace_fragments(params)
    params = unwrap_single(params)
    params = maybe_replace_order(fun, params)

    new_params = Qh.Expr.maybe_deep_prefix_binding(binding_provided, params)

    if !binding_provided and params == new_params do
      quote do
        query = Ecto.Query.unquote(fun)(query, unquote(new_params))
      end
    else
      quote do
        query = Ecto.Query.unquote(fun)(query, unquote(binding), unquote(new_params))
      end
    end
  end

  defp maybe_replace_order(:order_by, order) do
    Enum.map(List.wrap(order), fn
      {key, dir}
      when dir in [
             :asc,
             :asc_nulls_last,
             :asc_nulls_first,
             :desc,
             :desc_nulls_last,
             :desc_nulls_first
           ] ->
        {dir, key}

      other ->
        other
    end)
  end

  defp maybe_replace_order(_fun, params) do
    params
  end

  defp replace_fragments(list) do
    case list do
      [q | _] when is_binary(q) ->
        quote do: fragment(unquote_splicing(list))

      list ->
        list
    end
  end

  defp expr_to_funcs(expr) do
    case expr do
      list when is_list(list) -> Enum.map(list, fn sub_expr -> expr_to_funcs(sub_expr) end)
      {fun, arg} -> {fun, [], [{{:., [], [{:&, [], [0]}, arg]}, [], []}]}
      fun when is_atom(fun) -> {fun, [], []}
      expr -> expr
    end
  end

  #

  def has_group_by?(%{group_bys: list}) when length(list) > 0, do: true
  def has_group_by?(_query), do: false

  def default_primary_order(schema) when is_atom(schema) do
    require Ecto.Query
    primary_key = schema.__schema__(:primary_key)
    Ecto.Query.order_by(schema, ^primary_key)
  end

  def default_primary_order(%{from: %{source: {_, schema}}, order_bys: []} = query) do
    require Ecto.Query
    primary_key = schema.__schema__(:primary_key)
    Ecto.Query.order_by(query, ^primary_key)
  end

  def default_primary_order(unknown), do: unknown

  def get_schema(schema) when is_atom(schema) do
    schema
  end

  def get_schema(%{from: %{source: {_, schema}}}) do
    schema
  end

  def split_optional_binding(params) do
    if first_param_binding?(params) do
      [binding | params] = params
      {true, binding, params}
    else
      binding = quote do: [t]
      {false, binding, params}
    end
  end

  defp first_param_binding?([maybe_binding | _]), do: is_binding?(maybe_binding)
  defp first_param_binding?(_), do: false

  defp is_binding?(list) do
    is_list(list) &&
      Enum.all?(list, fn
        {key, _, nil} when is_atom(key) -> true
        {alias, {key, _, nil}} when is_atom(alias) and is_atom(key) -> true
        _other -> false
      end)
  end

  def repo(opts \\ []) do
    repo_mod(opts)
  end

  def repo_mod(opts) do
    opts[:repo] || Application.get_env(:qh, :repo) || Module.concat(app_mod(opts), Repo)
  end

  def app_mod(opts) do
    opts[:app_mod] || maybe_camelize(opts[:app]) || Application.get_env(:qh, :app_mod) ||
      maybe_camelize(Application.get_env(:qh, :app))
  end

  def lookup_schema(schema, opts \\ []) do
    app_mod = app_mod(opts)

    cond do
      is_atom(schema) and first_uppercase?(schema) and ensure_compiled?(schema) ->
        schema

      first_uppercase?(schema) ->
        module = Module.concat(app_mod, schema)
        module_or_raise!(module, schema)

      true ->
        schema
    end
  end

  defp module_or_raise!(module, schema) do
    if ensure_compiled?(module) do
      module
    else
      raise "unable to find schema for #{inspect(schema)}, #{inspect(module)} does not exist"
    end
  end

  defp ensure_compiled?(module) do
    case Code.ensure_compiled(module) do
      {:module, _} ->
        true

      {:error, _} ->
        false
    end
  end

  defp first_uppercase?(val) do
    String.at(to_string(val), 0) == String.capitalize(String.at(to_string(val), 0))
  end

  defp maybe_camelize(nil), do: nil
  defp maybe_camelize(val), do: Macro.camelize(to_string(val))


  def unwrap_single([item]), do: item
  def unwrap_single(other), do: other

  defp unwrap_nested({{:., _, [left, right]}, opts, children}) do
    parens = if opts[:no_parens], do: false, else: true

    unwrap_nested(left) ++ [{right, [parens: parens], children}]
  end

  defp unwrap_nested({k, _, nil}), do: [k]
  defp unwrap_nested(v), do: [v]
end
